server/src/main/java/org/opensearch/action/admin/cluster/decommission/put/TransportPutDecommissionAction.java

```diff
@@ -0,0 +1,248 @@
+            ActionListener<AddVotingConfigExclusionsResponse> addVotingConfigExclusionsListener = new ActionListener<>() {
+                @Override
+                public void onResponse(AddVotingConfigExclusionsResponse addVotingConfigExclusionsResponse) {
+                    logger.info("Master abdicated - Response received");
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+            };
+            exclusionsAction.execute(new AddVotingConfigExclusionsRequest(masterNode.getName()), addVotingConfigExclusionsListener);
+            throw new NotClusterManagerException("abdicated");
+        }
+
+
+        List<DiscoveryNode> decommissionedNodes = getDecommissionedNodes(request.getDecommissionAttribute(), state);
+        checkHttpStatsForDecomissionedNodes(decommissionedNodes, listener);
+
+        decommissionService.registerDecommissionAttribute(
+            request,
+            ActionListener.delegateFailure(
+                listener,
+                (delegatedListener, response) -> delegatedListener.onResponse(new PutDecommissionResponse(response.isAcknowledged()))
+            )
+        );
+    }
+
+    private void checkHttpStatsForDecomissionedNodes(List<DiscoveryNode> decommissionedNodes, ActionListener<PutDecommissionResponse> listener) {
+        ActionListener<NodesStatsResponse> nodesStatsResponseActionListener = new ActionListener<NodesStatsResponse>() {
+            @Override
+            public void onResponse(NodesStatsResponse nodesStatsResponse) {
+                boolean hasActiveConnections = false;
+                List<NodeStats> responseNodes = nodesStatsResponse.getNodes();
+                for (int i=0; i < responseNodes.size(); i++) {
+                    HttpStats httpStats = responseNodes.get(i).getHttp();
+                    if (httpStats != null && httpStats.getServerOpen() != 0) {
+                        hasActiveConnections = true;
+                        break;
+                    }
+                }
+                if (hasActiveConnections) {
+                    // Slow down the next call to get the Http stats from the decommissioned nodes.
+                    delay(1000L);
+                    // Check again for active connections. Repeat the process till we have no more active requests open.
+                    waitForGracefulDecommission(decommissionedNodes, this);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        };
+        waitForGracefulDecommission(decommissionedNodes, nodesStatsResponseActionListener);
+    }
+
+    private void delay(long delayMillis) {
+        try {
+            Thread.sleep(delayMillis);
+        } catch (InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            return;
+        }
+    }
+
```

Comment:
Please avoid Thread sleep, instead use a scheduler to update the drain state and invoke callbacks